src/Microsoft.AspNetCore.Razor.Runtime/TagHelpers/DefaultTagHelperContent.cs

```diff
@@ -31,16 +33,38 @@ private List<object> Buffer
                     _buffer = new List<object>();
                 }
 
+                if (_isSingleContentSet)
+                {
+                    Debug.Assert(_buffer.Count == 0);
+
+                    _buffer.Add(_singleContent);
+                    _isSingleContentSet = false;
+                }
+
                 return _buffer;
             }
         }
 
+        private object SingleContent
+        {
+            get
+            {
+                return _singleContent;
+            }
+            set
+            {
+                _isModified = true;
+                _isSingleContentSet = true;
+                _singleContent = value;
+            }
+        }
+
         /// <inheritdoc />
         public override bool IsModified => _isModified;
 
         /// <inheritdoc />
         /// <remarks>Returns <c>true</c> for a cleared <see cref="TagHelperContent"/>.</remarks>
-        public override bool IsWhiteSpace
+        public override bool IsEmptyOrWhiteSpace
```

Comment:
Why are changes to the public surface mixed into this PR? We decided long ago to support `IsEmpty` and I'm not really sure why the `<label>` and validation message tag helpers aren't using it these days.

In any case, it's hard to see what's actually changing due to the actual #621 fix. If you think this is valuable, please open an issue and submit separately.